### PR TITLE
420 execution metrics zio

### LINKED ISSUE #420: Auto-Instrument ZIO 2.0 execution metrics
[ZIO Executors](https://zio.dev/api/zio/internal/executionmetrics) expose some relevant metrics, such as: 
- workers_count,
- concurrency,
- executor_capacity,
- task_queue_size, 
- enqueued_count,
- dequeued_count

By auto-instrumenting the Executor's constructor, we can get a reference to it. Such a reference can be then used in asynchronous instrument callbacks to periodically collect execution metrics and report them via openTelemetry API. 

Proposal Link with code snippets on how to do that (see Part 1): [here](https://docs.google.com/document/d/18pR8oGcXXPyZF-L6CKch_ABzLRgtYJtFn-IHwrPCFx8/edit?usp=sharing)

Configuration: 
 - we should be able to turn the whole instrumentation on/off (that's for free thanks to openTelemetry)
 - nothing else should be configured for now in order to keep this ticket small and easy to review. In the future we can extend the config

Things to be ironed out: 
 - the proposal distinguishes the executors by calling executor.toString() which is a naive way of doing this. Is there a better way? 
 - is there a way to have an extra attribute: isBlocking (true/false). It seems beneficial to be able to distinguish blocking/non-blocking executors using attributes.

### PR DESCRIPTION
**Pull request by lgajowy**
Closes #420  
___

## Summary:

 - adds executor metrics
 - adds tests for the metrics

### CODE REVIEW

File: otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+object ZIOExecutorMetrics {
+
+  private val meter: Meter = GlobalOpenTelemetry.getMeter("mesmer")
+
+  private val executionMetricName: String => String = (suffix: String) => s"mesmer_zio_executor_$suffix"
+
+  def registerExecutorMetrics(executor: Executor): Unit =
+    Unsafe.unsafe { implicit u =>
+      val executorAttributes = Attributes.builder().put("executor", executor.toString).build()
```
Comment (worekleszczy) [inline otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala:22]: I think that we should try to distinguish between default `ZScheduler` and `zio.internal.Blocking.blockingExecutor` that should be consistent across multiple run of the application - right now if a service is restarted that run with default `Executor` `exector` attribute will change

File: otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+  private def registerAsyncLongGauge(name: String, unit: String, description: String)(
+    callback: Consumer[ObservableLongMeasurement]
+  ): ObservableLongGauge =
+    meter
+      .gaugeBuilder(name)
+      .ofLongs()
+      .setUnit(unit)
+      .setDescription(description)
+      .buildWithCallback(callback)
```
Comment (worekleszczy) [inline otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala:73]: To protect resources we should add a deregistration logic on executor shutdown

File: otel-extension/src/it/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetricsTest.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.util.concurrent.{ SynchronousQueue, ThreadPoolExecutor, TimeUnit }
+import scala.jdk.CollectionConverters._
+
+class ZIOExecutorMetricsTest extends OtelAgentTest with AnyFlatSpecLike with Matchers {
+
+  val testProgram: ZIO[Any, Nothing, Long] = (for {
+    _ <- Random.nextInt
+  } yield ())
+    .repeat(Schedule.forever)
```
Comment (worekleszczy) [inline otel-extension/src/it/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetricsTest.scala:18]: Maybe just 5 sec will be ok?

Comment (lgajowy) [reply otel-extension/src/it/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetricsTest.scala:18]: Wouldn't it potentially be flaky?

File: otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala (hunk 1 above)
Comment (lgajowy) [reply otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala:22]: Good point. Do you think config parameters will be a good idea?

Comment (lgajowy) [reply otel-extension/src/main/scala/io/scalac/mesmer/otelextension/instrumentations/zio/ZIOExecutorMetrics.scala:22]: But, the question is how do we distinguish the blocking/non-blocking executor in the advice?

**Review by worekleszczy — APPROVED** [review]

**Pull request merged**